includes/Checker/Check_Result.php

```diff
@@ -7,11 +7,204 @@
 
 namespace WordPress\Plugin_Check\Checker;
 
+use WordPress\Plugin_Check\Plugin_Context;
+use WordPress\Plugin_Check\Checker\Check_Context;
+
 /**
  * Result for running checks on a plugin.
  *
  * @since n.e.x.t
  */
 class Check_Result {
-	// @TODO: Complete as part of issue #11.
+
+	/**
+	 * Main context instance.
+	 *
+	 * @since n.e.x.t
+	 * @var Plugin_Context
+	 */
+	protected $main_context;
+
+	/**
+	 * Context for the plugin to check.
+	 *
+	 * @since n.e.x.t
+	 * @var Check_Context
+	 */
+	protected $check_context;
+
+	/**
+	 * List of errors.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $errors = array();
+
+	/**
+	 * List of warnings.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	protected $warnings = array();
+
+	/**
+	 * Number of errors.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	protected $error_count = 0;
+
+	/**
+	 * Number of warnings.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	protected $warning_count = 0;
+
+	/**
+	 * Sets the context for the plugin to check.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Plugin_Context $main_context  Main context instance.
+	 * @param Check_Context  $check_context Check context instance for the plugin.
+	 */
+	public function __construct( Plugin_Context $main_context, Check_Context $check_context ) {
+		$this->main_context  = $main_context;
+		$this->check_context = $check_context;
+	}
+
+	/**
+	 * Returns the main context for the plugin checker.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return Plugin_Context Main context instance.
+	 */
+	public function context() {
+		return $this->main_context;
+	}
+
+	/**
+	 * Returns the context for the plugin to check.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return Check_Context Plugin context instance.
+	 */
+	public function plugin() {
+		return $this->check_context;
+	}
+
+	/**
+	 * Adds an error or warning to the respective stack.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param bool   $error   Whether it is an error message.
+	 * @param string $message The message.
+	 * @param array  $args    {
+	 *     Additional message arguments.
+	 *
+	 *     @type string $code   Violation code according to the message. Default empty string.
+	 *     @type string $file   The file in which the message occurred. Default empty string (unknown file).
+	 *     @type int    $line   The line on which the message occurred. Default 0 (unknown line).
+	 *     @type int    $column The column on which the message occurred. Default 0 (unknown column).
+	 * }
+	 */
+	public function add_message( $error, $message, $args = array() ) {
+		$defaults = array(
+			'code'   => '',
+			'file'   => '',
+			'line'   => 0,
+			'column' => 0,
+		);
+
+		$data = array_merge(
+			array(
+				'message' => $message,
+			),
+			$defaults,
+			array_intersect_key( $args, $defaults )
+		);
+
+		$file   = str_replace( $this->check_context->path( '/' ), '', $data['file'] );
+		$line   = $data['line'];
+		$column = $data['column'];
+		unset( $data['line'], $data['column'] );
```

Comment:
Should we unset `$data['file']` here too (since it's already in the structure below, like line and column are)?